### PR TITLE
Enhancement/write fit file

### PR DESCRIPTION
**Pull request by mgmart**
fit files are successfully read by Rubitrack 

Some more testing would be appreciated ;)

### CODE REVIEW

File: GpxExport/ViewControllers/WorkoutsTableViewController.swift [hunk 1]
```diff
@@ -90,6 +90,19 @@ class WorkoutsTableViewController: UITableViewController {
     }
 
     private func exportSelectedWorkouts() {
+        let alert = UIAlertController(title: "What file to generate?", message: "It's recommended to use GPX for now.", preferredStyle: .actionSheet)
```
Comment (patricks) [inline GpxExport/ViewControllers/WorkoutsTableViewController.swift]: The strings should be localized.

File: GpxExport/ViewControllers/WorkoutsTableViewController.swift [hunk 2]
```diff
@@ -90,6 +90,19 @@ class WorkoutsTableViewController: UITableViewController {
...
+
+        alert.addAction(UIAlertAction(title: "GPX", style: .default, handler: {(_: UIAlertAction!) in
+            self.handleExport("GPX")
```
Comment (patricks) [inline GpxExport/ViewControllers/WorkoutsTableViewController.swift]: Would it make sense to create something like this

```swift
enum ExportFileType {
    case gpx
    case fit
}
```

File: GpxExport/ViewControllers/WorkoutsTableViewController.swift [hunk 3]
```diff
@@ -90,6 +90,19 @@ class WorkoutsTableViewController: UITableViewController {
     }
 
     private func exportSelectedWorkouts() {
+        let alert = UIAlertController(title: "What file to generate?", message: "It's recommended to use GPX for now.", preferredStyle: .actionSheet)
+
+        alert.addAction(UIAlertAction(title: "GPX", style: .default, handler: {(_: UIAlertAction!) in
```
Comment (patricks) [inline GpxExport/ViewControllers/WorkoutsTableViewController.swift]: you cloud use the handler without force unwrapping.

File: GpxExport/ViewControllers/WorkoutsTableViewController.swift (hunk 1 above)
Comment (mgmart) [reply GpxExport/ViewControllers/WorkoutsTableViewController.swift]: Yeah, of course. I forgot that we do that now ;)

File: GpxExport/ViewControllers/WorkoutsTableViewController.swift (hunk 2 above)
Comment (mgmart) [reply GpxExport/ViewControllers/WorkoutsTableViewController.swift]: Yeah, lazy me ...

File: GpxExport/ViewControllers/WorkoutsTableViewController.swift (hunk 3 above)
Comment (mgmart) [reply GpxExport/ViewControllers/WorkoutsTableViewController.swift]: Oops. Maybe an copy/paste issue

File: GpxExport/Models/Workout+FitWriter.swift [hunk 4]
```diff
@@ -42,74 +42,80 @@ extension Workout {
...
+                                           eventGroup: nil)
+            messages.append(activity)
+            let fileId = FileIdMessage(deviceSerialNumber: nil,
+                                        fileCreationDate: time,
+                                        manufacturer: .garmin,
+                                        product: nil,
+                                        fileNumber: nil,
+                                        fileType: FileType.activity,
+                                        productName: nil)
+
```
Comment (khoogheem) [inline GpxExport/Models/Workout+FitWriter.swift]: Couple of things I would recommend.
Always have a Serial Number even it if is 1 same thing for File Number.

File: GpxExport/Models/Workout+FitWriter.swift [hunk 5]
```diff
@@ -42,74 +42,80 @@ extension Workout {
...
-            file.write("""
-  </trkseg>
-  </trk>
-  </gpx>
 
-  """.data(using: .utf8)!)
-            file.closeFile()
+            do {
+                let encoder = FitFileEncoder(dataValidityStrategy: .none)
+                let data = try encoder.encode(fildIdMessage: fileId, messages: messages)
```
Comment (khoogheem) [inline GpxExport/Models/Workout+FitWriter.swift]: I would turn on the ValidityStrategy to FileType or GarminConnect.  Use the garminConnect option if you ever expect to send it up to GarminConnect in the Future, as they have some other Rules applied to Activity File type that are stricter then the standard Required fields for activity type files

Comment (khoogheem) [reply GpxExport/Models/Workout+FitWriter.swift]: Along with that you should probably add in a DeviceInfoMessage.

**Comment by khoogheem** [comment]
Just left a few comments that will hopefully help as you test out FIT files on sites like Strava/GarminConnect, etc.  I highly recommend that you turn on the validity checks as it will make sure your ActivityFile format is correct

**Comment by mgmart** [comment]
Thanks @khoogheem. At the moment Strava does indeed accept the generated files. Garmin Connect does not, of course. 

Will incorporate your suggestions. Then hopefully also GC will accept the files.

**Comment by khoogheem** [comment]
Yep.. we struggled with GC for a while.. the gc Validations will help get that working.  Strava was pretty much plug and play.

**Comment by khoogheem** [comment]
Here is generically the way I have used the encoding:
```
    private class func fitencode(records: [WorkoutRecord]) throws -> Data {

        let fileDate = FitTime(date: Date())

        var productID: ValidatedBinaryInteger<UInt16> {

            if let record = records.first {
                return ValidatedBinaryInteger(value: record.data.getDisplayType().rawValue, valid: true)
            }

            return ValidatedBinaryInteger(value: GymMachineType.invalid.rawValue, valid: true)
        }

        let fileIDMessage = FileIdMessage(deviceSerialNumber: ValidatedBinaryInteger(value: UInt32(1), valid: true),
                                          fileCreationDate: fileDate,
                                          manufacturer: Manufacturer.northPoleEngineering,
                                          product: productID,
                                          fileNumber: ValidatedBinaryInteger(value: UInt16(1), valid: true),
                                          fileType: FileType.activity,
                                          productName: nil)

        let deviceInfoMessage = createDeviceInfoMessage(fileDate: fileDate, productID: productID)
        let recordMessages = records.compactMap({ $0.data.toRecordMessage() })

        var fitMessages: [FitMessage] = [deviceInfoMessage] + recordMessages

        let machineData = records.compactMap({ $0.data })

/// Create One Sport Message 
        let sport = createSportMesssage(records: machineData)
        fitMessages += [sport]

/// This creates a One Lap and One Session Message
        let combo = createLapSessionMessage(records: machineData)
        fitMessages += [combo.lapMessage]
        fitMessages += [combo.sessionMessage]

        let activity = createActivityMessage(fileDate: fileDate, duration: combo.sessionMessage.totalTimerTime)
        fitMessages += [activity]

        do {
            let encoder = FitFileEncoder(dataValidityStrategy: .garminConnect)

            let fileData = try encoder.encode(fildIdMessage: fileIDMessage, messages: fitMessages)
            return fileData

        } catch {
            EventLogger.common.localLog(error)
            throw ApplicationError(message: "\(error)")
        }
    }
```

**Comment by mgmart** [comment]
Thanks again. That will help very much!

File: GpxExport/Models/Workout+FitWriter.swift [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            let file: FileHandle
+
+            do {
+                let manager = FileManager.default
+                if manager.fileExists(atPath: targetURL.path) {
+                    try manager.removeItem(atPath: targetURL.path)
+                }
+                manager.createFile(atPath: targetURL.path, contents: Data())
+                file = try FileHandle(forWritingTo: targetURL)
+            } catch let err {
```
Comment (patricks) [inline GpxExport/Models/Workout+FitWriter.swift:41]: Btw. you can also access error without the `let`.

```
} catch {
 print(error)
}
```

I also would output `error.debugDescription` or `error.localizedDescription`.

**Review by patricks — APPROVED** [review]
Nice improvement 👍

**Pull request merged**